### DISCUSSION TITLE
Developing on triton server

### DISCUSSION THREAD

**Discussion by coolkp** [triton-inference-server/server#7782] [General]
Hi folks,
Happy friday. I am currently customising triton server to add some meta information to response headers. My changes are only in server, I only want to build server. I want to build my tritonserver and deploy with tensor rt llm backend and llama 3 model, Whats the best most efficient way to do this. Should i make my changes then run compose with `min` image of `trtllm-python-py3`. How do I use this image built by compose then to run my inference

**Comment by coolkp** [comment]
Building from source everything takes super long. So i think i am not doing this right

**Reply by coolkp** [reply]
@yinggeh if you have some context, would appreciate the help

**Reply by yinggeh** [reply]
@jbkyang-nvi  is probably the most knowledgeable in `compose.py`.
Hi @jbkyang-nvi, can you help?

**Reply by coolkp** [reply]
also is there a bug in build.py, cmake is trying to clone https://github.com/triton-inference-server/tensorrt_llm_backend.git which doesn't exist. https://github.com/triton-inference-server/tensorrtllm_backend.git does exist but is missing CMakeLists.txt?

**Reply by coolkp** [reply]
@jbkyang-nvi

**Reply by jbkyang-nvi** [reply]
> also is there a bug in build.py, cmake is trying to clone https://github.com/triton-inference-server/tensorrt_llm_backend.git which doesn't exist. https://github.com/triton-inference-server/tensorrtllm_backend.git does exist but is missing CMakeLists.txt?

can you point me to where this is? I don't see any reference to tensorrt_llm_backend in build.py

**Reply by coolkp** [reply]
@jbkyang-nvi thanks for the reply. Ok looks like passing in `--backend= tensorrtllm` avoids that error, since repo link is constructed based of the name. But ran into another issue.  Also i am not sure if this is the right way to develop on server. Is there a guide to develop only build the server code changes and compose with the tensorrtllm backend
![image](https://github.com/user-attachments/assets/64f3e0b5-fca3-48f7-b4ee-06f831dcf538)

**Reply by krishung5** [reply]
@jbkyang-nvi to correct me if I'm wrong. I believe the compose.py doesn't support tensorrtllm_backend at the moment.

We have slightly different build steps for building TRT-LLM container. Following the build instructions [here](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/docs/build.md#option-1-build-the-ngc-triton-trt-llm-container), inside the [`build.sh`](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/build.sh#L40) script you can see that we are building a trtllm_base image first, which builds and installs the TRT-LLM library, then we run `build.py` to build Triton sever and the backends with that trtllm_base image as base image. I think in this case, if you want to only rebuild the Triton server, you could try something like this:
```
# Inside your server repo
TRTLLM_BASE_IMAGE=nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3
./build.py -v --no-container-interactive --enable-logging --enable-stats --enable-tracing \
              --enable-metrics --enable-gpu-metrics --enable-cpu-metrics \
              --filesystem=gcs --filesystem=s3 --filesystem=azure_storage \
              --endpoint=http --endpoint=grpc --endpoint=sagemaker --endpoint=vertex-ai \
              --backend=ensemble --enable-gpu --no-container-pull \
              --repoagent=checksum --cache=local --cache=redis \
              --image=base,${TRTLLM_BASE_IMAGE}
``` 

UPDATE:
The `tritonserver` image built from above will not include the TRT-LLM backend. You'd have to replace the Triton bits in `nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3` with the ones in the `tritonserver` image: `/opt/tritonserver/lib` and `/opt/tritonserver/bin`.

**Reply by coolkp** [reply]
Yes, it doesn't support but documentation says that you can use `--image min,nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3` does this method build server with local changes and tensortllm backend?

**Reply by krishung5** [reply]
I'm not sure about the compose.py part, I think you could see the detailed steps from the dockerfile with `--dry-run` enabled. I assume it should be building from your local server repo if you running compose.py inside your repo.

For build.py, if you running into the local server repo, then yes it will be building with your local changes. The command I shared doesn't include the flag `--backend=tensorrtllm:${TENSORRTLLM_BACKEND_REPO_TAG} `, so it will not build the tensorrtllm backend. If you want to build the backend as well, you' need to append the flag when running build.py.

**Reply by krishung5** [reply]
@coolkp Updated my previous comment above. I think you'd have to replace Triton bits in the TRT-LLM container with the one you just build.

**Reply by coolkp** [reply]
Where is the dockerfile, TRT-LLM container, i assume you mean, TRT-LLM container is `nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3` ?  I understand this has trtllm engine+backend with python. And used as base/min image in compose it should be present in backends? but i don't see it.  There is step to copy nto sure where to copy .   I believe `COPY --chown=1000:1000 --from=full /opt/tritonserver/backends/pytorch /opt/tritonserver/backends/pytorch` should be changed?
```Dockerfile
#
# Multistage build.
#
ARG TRITON_VERSION=2.51.0
ARG TRITON_CONTAINER_VERSION=24.10

FROM nvcr.io/nvidia/tritonserver:24.10-py3 AS full

FROM nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3

ARG TRITON_VERSION
ARG TRITON_CONTAINER_VERSION

ENV TRITON_SERVER_VERSION ${TRITON_VERSION}
ENV NVIDIA_TRITON_SERVER_VERSION ${TRITON_CONTAINER_VERSION}
LABEL com.nvidia.tritonserver.version="${TRITON_SERVER_VERSION}"

ENV PATH /opt/tritonserver/bin:${PATH}
# Remove once https://github.com/openucx/ucx/pull/9148 is available
# in the min container.
ENV UCX_MEM_EVENTS no

ENV LD_LIBRARY_PATH /opt/hpcx/ucc/lib/:/opt/hpcx/ucx/lib/:${LD_LIBRARY_PATH}

ENV TF_ADJUST_HUE_FUSED         1
ENV TF_ADJUST_SATURATION_FUSED  1
ENV TF_ENABLE_WINOGRAD_NONFUSED 1
ENV TF_AUTOTUNE_THRESHOLD       2
ENV TRITON_SERVER_GPU_ENABLED    1

# Create a user that can be used to run triton as
# non-root. Make sure that this user to given ID 1000. All server
# artifacts copied below are assign to this user.
ENV TRITON_SERVER_USER=triton-server
RUN userdel tensorrt-server > /dev/null 2>&1 || true \
      && if ! id -u $TRITON_SERVER_USER > /dev/null 2>&1 ; then \
          useradd $TRITON_SERVER_USER; \
        fi \
      && [ `id -u $TRITON_SERVER_USER` -eq 1000 ] \
      && [ `id -g $TRITON_SERVER_USER` -eq 1000 ]

# Ensure apt-get won't prompt for selecting options
ENV DEBIAN_FRONTEND=noninteractive

# Common dependencies. FIXME (can any of these be conditional? For
# example libcurl only needed for GCS?)
RUN apt-get update \
      && apt-get install -y --no-install-recommends \
              clang \
              curl \
              dirmngr \
              git \
              gperf \
              libb64-0d \
              libcurl4-openssl-dev \
              libgoogle-perftools-dev \
              libjemalloc-dev \
              libnuma-dev \
              libre2-9 \
              software-properties-common \
              wget \
              libgomp1 \
      && rm -rf /var/lib/apt/lists/*

# Set TCMALLOC_RELEASE_RATE for users setting LD_PRELOAD with tcmalloc
ENV TCMALLOC_RELEASE_RATE 200

ENV DCGM_VERSION 3.2.6
# Install DCGM. Steps from https://developer.nvidia.com/dcgm#Downloads
RUN curl -o /tmp/cuda-keyring.deb \
          https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64/cuda-keyring_1.0-1_all.deb \
      && apt install /tmp/cuda-keyring.deb \
      && rm /tmp/cuda-keyring.deb \
      && apt-get update \
      && apt-get install -y datacenter-gpu-manager=1:3.2.6

# Extra defensive wiring for CUDA Compat lib
RUN ln -sf ${_CUDA_COMPAT_PATH}/lib.real ${_CUDA_COMPAT_PATH}/lib \
    && echo ${_CUDA_COMPAT_PATH}/lib > /etc/ld.so.conf.d/00-cuda-compat.conf \
    && ldconfig \
    && rm -f ${_CUDA_COMPAT_PATH}/lib

WORKDIR /opt/tritonserver
RUN rm -fr /opt/tritonserver/*
ENV NVIDIA_PRODUCT_NAME="Triton Server"
COPY docker/entrypoint.d/ /opt/nvidia/entrypoint.d/

ENV NVIDIA_BUILD_ID 117849258
LABEL com.nvidia.build.id=117849258
LABEL com.nvidia.build.ref=cd309d79dc2130ab890fd63def83fa8c17668393

WORKDIR /opt/tritonserver
COPY --chown=1000:1000 --from=full /opt/tritonserver/LICENSE .
COPY --chown=1000:1000 --from=full /opt/tritonserver/TRITON_VERSION .
COPY --chown=1000:1000 --from=full /opt/tritonserver/NVIDIA_Deep_Learning_Container_License.pdf .
COPY --chown=1000:1000 --from=full /opt/tritonserver/bin bin/
COPY --chown=1000:1000 --from=full /opt/tritonserver/lib lib/
COPY --chown=1000:1000 --from=full /opt/tritonserver/include include/
# Copying over backends 
COPY --chown=1000:1000 --from=full /opt/tritonserver/backends/pytorch /opt/tritonserver/backends/pytorch

# Top-level /opt/tritonserver/backends not copied so need to explicitly set permissions here
RUN chown triton-server:triton-server /opt/tritonserver/backends
#  Copying over repoagents 
#  Copying over caches 

LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
COPY --chown=1000:1000 --from=full /usr/bin/serve /usr/bin/.
```

**Reply by coolkp** [reply]
@krishung5

**Reply by krishung5** [reply]
> And used as base/min image in compose it should be present in backends? but i don't see it

As mentioned above, compose.py doesn't support TRT-LLM backend, so I doubt using this way would work.

> I believe COPY --chown=1000:1000 --from=full /opt/tritonserver/backends/pytorch /opt/tritonserver/backends/pytorch should be changed

I think modify this line to copy the TRT-LLM backend should work. What I was suggesting was building a new container with your server changes, and copy the server libraries into the TRT-LLM container `nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3`.

**Reply by coolkp** [reply]
how to do this

>  What I was suggesting was building a new container with your server changes

and where to copy @krishung5 

> copy the server libraries into the TRT-LLM container

**Reply by krishung5** [reply]
@coolkp Please see my above [comment](https://github.com/triton-inference-server/server/discussions/7782#discussioncomment-11194253). After building the server with build.py command, you will get an image named `tritonserver`. The server libraries are located in `/opt/tritonserver/lib` and `/opt/tritonserver/bin`. Just replace them in the TRT-LLM container `nvcr.io/nvidia/tritonserver:24.10-trtllm-python-py3` should work.

**Comment by jeetwin-az** [comment]
Hi!

To efficiently build and deploy your customized Triton server with the TensorRT LLM backend and Llama 3 model, here's a streamlined approach:

1. **Customize Triton Server**: Make your changes to the server code to add meta information to response headers.

2. **Build the Server**: Use `docker-compose` with the `tritonserver` base image to build your customized Triton server.

3. **Use TensorRT LLM Backend**: In your `docker-compose.yml`, specify the `trtllm-python-py3` image as the base for the TensorRT LLM backend, which will handle inference.

4. **Deploy**: After building with `docker-compose`, you can run the inference by starting the server container and pointing it to your model and configuration files.

5. **Run Inference**: Once the Triton server is up, use the Triton client or API to send inference requests to the server.

Let me know if you need further details on any specific step!
